### ISSUE TITLE
Feature: Closest Road to a certain location's coordinates

### ISSUE THREAD

**Issue by miskolc** [gboeing/osmnx#198]
I'm using osmnx and networkx for calculating shortest routes in Dubai and I'm having an issue when finding the closest road to a tracking location's coordinates. Even if I use coordinates that are right on a specific highway, sometimes it is matched to a close road due to the fact that I use the `osmnx.get_nearest_node` to find the closest road. This function will search for the closest graph node in the OSM map to that specific location. Due to the fact that for certain highways you have very large distances between consective node, the closest node might not be on the same road on which the  tracking location was recorded. It would be nice if there was a method `osmnx.get_nearest_edge` that mapped each location to the closest road.

Here is an example with 2 orig_coordinates that were recorded on the same highway but the second one is matched to a node on a different road.
I recommend using jupyter notebooks to visualize the problem on the map

```python
import networkx as nx
import osmnx as ox
import requests
from matplotlib import pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors
ox.config(use_cache=True, log_console=True)
ox.__version__

# get a graph for some city
SMALL_MARINA = [25.09, 25.06, 55.16, 55.11]
G2 = ox.graph_from_bbox(
    *SMALL_MARINA,
    simplify=False,
    retain_all=True,
    network_type='drive',

)
fig2, ax2 = ox.plot_graph(G2, fig_height=20, fig_width=20)

# get the nearest network node to each point

good_orig_node = ox.get_nearest_node(G2, (25.070661, 55.137911), method='euclidean')
bad_orig_node = ox.get_nearest_node(G2, (25.071764, 55.138978), method='euclidean')
dest_node = ox.get_nearest_node(G2, (25.079223, 55.146241), method='euclidean')



# find the route between these nodes then plot it
route = nx.shortest_path(G2, good_orig_node, dest_node, weight='length')
fig, ax = ox.plot_graph_route(G2, route, fig_height=20, fig_width=20)


# find the route between these nodes then plot it
route = nx.shortest_path(G2, bad_orig_node, dest_node, weight='length')
fig, ax = ox.plot_graph_route(G2, route, fig_height=20, fig_width=20)
```
### Coordinate(South-West) is matched to the right road
<img width="1006" alt="screen shot 2018-08-24 at 17 07 42" src="https://user-images.githubusercontent.com/2085350/44595220-4ea81d80-a7c0-11e8-90a6-61ba02d9578a.png">

### Coordinate(South-West) is matched to the wrong road

<img width="1040" alt="screen shot 2018-08-24 at 17 07 59" src="https://user-images.githubusercontent.com/2085350/44595221-4ea81d80-a7c0-11e8-83bf-fb22f67e1e6c.png">

**Comment by RunOrVeith** [comment]
I guess this would be a good starting point https://blog.mapbox.com/matching-gps-traces-to-a-map-73730197d0e2

**Comment by MustafaIsmaill** [comment]
@miskolc i have the same exact issue, have you managed to find a workaround or a solution for this? would really appreciate your reply.

**Comment by RunOrVeith** [comment]
I'm running a docker container [serving osrm](https://hub.docker.com/r/osrm/osrm-backend/) to do this. It works well, but requires a bunch of boilerplate code.

**Comment by MustafaIsmaill** [comment]
@RunOrVeith  thanks! that's really helpful

**Comment by miskolc** [comment]
@gboeing sorry I haven't been following this thread for a few month. Yes, I've implemented my own solution for my current company where I calculate the closest road to a coordinate and then pick the closest point on that road.

**Comment by MustafaIsmaill** [comment]
@miskolc can you clarify how did you manage to find the nearest road to a point or a coordinate without looking for the nearest node first?

**Comment by miskolc** [comment]
@MustafaIsmaill I turn the OSM network graph into a GeoPandas DataFrame containing all roads in the regions I'm interested. Each row contains the road coordinates and from that It's simple computational geometry. I calculate the distance from the current point to the edges then pick the ones that is the closest.
There are a few catches though as initially the approach described above gave the correct results in > 3 seconds. We work with millions of tracking location so having to wait 1 second to be able to match one GPS tracking to a road would be unacceptable. I managed to perform a few tricks that ended up speeding the processing x1000:
1) You would have to cache the gdf of the graph and perform the search on the gdf without touching the networkx graph. You need the gdf anyway and converting the graph to a gdf for every coordinate would be an expensive operation. Also convert the road coordinates to a numpy array for some operations.
2) I try to pick only roads who's coordinates are close to my GPS trackings. If you work with a region with an area of 100km by 100km, it would suffice to look at roads whose edges lie within 10km from your GPS location.
3) Perform the search above using BallTrees instead of an exhaustive O(N) search for the closest. If you would do a condinational search on a numpy array it will look through all the road coordinates in your area( ~1.000.000 for the map of a large city). However if you index these coordinates using a BallTree your searches will be performed in O(logN) time

**Comment by MustafaIsmaill** [comment]
@miskolc thank you for the info and great job!

**Comment by gboeing** [comment]
@miskolc do you have a gist or demonstration of your methodology? Is this something you'd be comfortable contributing to OSMnx as a pull request?

**Comment by miskolc** [comment]
@gboeing I created a small PR with a functional method that finds the nearest road to a point. Now again this is not very efficient as you need to get the GeoDataFrame from the Graph every time you query for a coordinate. If you would want that functionality to be included in a future version of OSMnx please let me know. Maybe it would be better to talk via email.

**Comment by samuelduchesne** [comment]
@miskolc and @gboeing, I had to calculate 100k+ nearest edges for one of my projects and I created a custom function that implements the ckdtree algorithm. When I stumbled upon this issue, I decided to give it a try and propose my solution.

I kept @miskolc's proposal as a fallback when the *method* name is not given. You will see that @miskolc's method is indeed much slower because the GeoDataFrame has to be generated at each call.

**Comment by akshayah3** [comment]
@samuelduchesne I have a very similar use case of matching a GPS trail to the corresponding edge of the graph. But in my case the GPS is inaccurate by a few meters and which results in the point being on the other side of the road. Then if I use the nearest node/edge it would give me the road on the other side of my actual road. I was thinking of calculating the bearing of my origin and destination and match it with the bearing of the closest edges/roads. Do you have any suggestions as to how could I approach this problem?

**Comment by samuelduchesne** [comment]
@akshayah3, what Do you mean by “the other side of the road”? Can you illustrate it?

**Comment by MustafaIsmaill** [comment]
@samuelduchesne I think you have to implement the hidden markov model in your code which includes finding the nearest nodes/edges. It also allows you to calculate a probability for the transition between roads/edges for which if gps point P(t) has been matched to road A, the probability of gps point P(t+1) being matched to road B would be low even if it is the closest, this is because changing from road A to B would be impossible because it is the opposite driving direction for example. You can find better explanation in the blog mentioned here before https://blog.mapbox.com/matching-gps-traces-to-a-map-73730197d0e2 
I found an easier way to do this, since I already know the path in which the vehicle will be travelling. I create gdf nodes and edges only from this path to avoid matching the points to other roads/edges and then I follow a similar solution to what is mentioned above to find the nearest edge for each node yet I don't know if this would be fit in your case or not.

**Comment by miskolc** [comment]
Indeed an HMMs is the most common to solve the problem of mapping a stream of GPS tracking locations from a producer to their actual coordinates. Uber has a nice [presentation](https://www.youtube.com/watch?v=ChtumoDfZXI) on how they are doing it.  

However this assumes you are actually working with multiple coordinates from the produced by the same entity (device/user). If you just want to match one single location to it's closest road it doesn't even make sense to use an HMM.

**Comment by gboeing** [comment]
Thanks everyone. Just to summarize, we have two PRs open. #231 and then #234 which builds on the former. They're in code review, and I'll merge them when that's done. For now, please take a look at them and share any relevant comments. Thanks!

**Comment by gboeing** [comment]
Closed in #231 and #234.

**Comment by ahmedtiko96** [comment]
can you make the get_nearest_edge() function available in the osmnx pip install?
I can already use it perfectly as i updated to my local utils.py file but it will be helpful for my team so that they have it directly when they install it using pip.
Thank you!

**Comment by gboeing** [comment]
@ahmedtiko96 this will be included soon in the next release, after #240 is also merged

**Comment by gboeing** [comment]
This has been released in v0.9

**Comment by SujitPattanayak** [comment]
problem statement:
1.why my route is going out of its destination point, i can't understand how to solve it. its is a very very big problem for me . plz help me @gboeing
![image](https://user-images.githubusercontent.com/97909067/177243679-c5a6821d-e060-4918-b367-e94cc6594a10.png)
